### ISSUE TITLE
magicgui layer list not updated when layer is renamed

### ISSUE THREAD

**Issue by dstansby** [napari/napari#4331]
## 🐛 Bug
The layer list in a widget created via. `magicgui` is not updated when a layer is renamed. It is correctly updated when a layer is added to the `napari` layers list, so I presume this bug is something missing in `magicgui` integration.

## To Reproduce

1. Open a `magicgui` created plugin with a `napari.layers.Image` widget
2. Open a new image in `napari`. This should update the image widget to show the name of the new image
3. Rename the image.  The image name does not update in the widget

## Expected behavior
The image name updates in the widget

## Environment
napari: 0.4.16.dev23+g9355fdbe
Platform: macOS-12.3-arm64-arm-64bit
System: MacOS 12.3
Python: 3.9.12 | packaged by conda-forge | (main, Mar 24 2022, 23:24:38) [Clang 12.0.1 ]
Qt: 5.15.2
PyQt5: 5.15.4
NumPy: 1.21.5
SciPy: 1.8.0
Dask: 2022.03.0
VisPy: 0.9.6

OpenGL:
- GL version: 2.1 Metal - 76.3
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1920x1080, scale 1.0
- screen 2: resolution 1440x900, scale 2.0

Plugins:
- brainreg-segment: 0.2.14
- console: 0.0.4
- napari-btrack: 0.1.dev3+gb039f8e
- ndtiffs: 0.1.2
- scikit-image: 0.4.16.dev23+g9355fdbe
- svg: 0.1.5

**Comment by psobolewskiPhD** [comment]
Eep, this one fell through the cracks huh.
I can confirm that renaming layers in the layer list does not update the layer combo box in a magicgui widget. Adding/removing a layer does though.
There is a `layer.events.name` event, so probably should listen for that?
I'm not actually sure where the layer list events are connected to magicgui things...

**Comment by DragaDoncila** [comment]
@psobolewskiPhD yeah, `layer.events.name` would be the one to watch out for. It might be quite tricky to hook it up correctly, though. Layerlist stuff gets hooked up to magicgui **very** tenuously [here](https://github.com/napari/napari/blob/main/napari/_qt/qt_main_window.py#L1177), simply by assuming that if a `widget` has a `reset_choices` attribute, we should call it whenever the layerlist changes.

Here, in theory, is where you could hook up to `layer.events.name` as well - the tricky bit is that because the `name` event is on the individual layer **not** on the layerlist, you'd need a two step approach of some sort e.g. a callback that connects to layerlist events which detects when a new layer is added, and then adds a callback to that layer's `name` event that resets choices on the widget?

We could also add a layerlist `name_changed` event that listens to individual `layer.name` events and re-emits them? This second one seems a bit easier... Then you could hook that up in the same spot.
